sync-core/src/main/java/com/cloudant/sync/query/QueryResult.java

```diff
@@ -55,109 +54,132 @@ public QueryResult(List<String> docIds,
-                }
+    private class QueryResultIterator implements Iterator<BasicDocumentRevision> {
+
+        private Iterator<BasicDocumentRevision> documentBlock;
+        private Range range;
+        private int nSkipped;
+        private int nReturned;
+
+        private QueryResultIterator() {
+            range = new Range(0, DEFAULT_BATCH_SIZE);
+            nSkipped = 0;
+            nReturned = 0;
+            documentBlock = populateDocumentBlock();
+        }
+
+        @Override
+        public boolean hasNext() {
+            return (documentBlock.hasNext() || range.location < originalDocIds.size());
+        }
+
+        @Override
+        public BasicDocumentRevision next() {
+            if (!this.hasNext()) {
+                throw new NoSuchElementException();
             }
-
-            @Override
-            public BasicDocumentRevision next() {
-                if(subIterator == null || !subIterator.hasNext()) {
-                    List<String> ids = subLists.remove(0);
-                    subIterator = this.nextSubIterator(ids);
-                }
-                
-                return subIterator.next();
+            if (!documentBlock.hasNext()) {
+                documentBlock = populateDocumentBlock();
             }
+            return documentBlock.next();
+        }
+
+        @Override
+        public void remove() {
+            throw new UnsupportedOperationException();
+        }
+
+        private Iterator<BasicDocumentRevision> populateDocumentBlock() {
+            List<BasicDocumentRevision> docList = new ArrayList<BasicDocumentRevision>();
+            while (range.location < originalDocIds.size()) {
+                boolean stop = false;
+                range.length = Math.min(DEFAULT_BATCH_SIZE, originalDocIds.size() - range.location);
+                List<String> batch = originalDocIds.subList(range.location,
+                                                            range.location + range.length);
+                List<BasicDocumentRevision> docs = datastore.getDocumentsWithIds(batch);
+                for (BasicDocumentRevision rev : docs) {
+                    BasicDocumentRevision innerRev;
+                    innerRev = rev;  // Allows us to replace later if projecting
+
+                    // Apply post-hoc matcher
+                    if (matcher != null && !matcher.matches(innerRev)) {
+                        continue;
+                    }
 
-            @Override
-            public void remove() {
-                throw new UnsupportedOperationException();
-            }
+                    // Apply skip (skip == 0 means disable)
+                    if (skip > 0 && nSkipped < skip) {
+                        nSkipped = nSkipped + 1;
+                        continue;
+                    }
+
+                    // Apply projection if result matches
+                    if (fields != null && !fields.isEmpty()) {
+                        // TODO - projection
```

Comment:
I believe that the resolution we came to today was to leave this PR as it is wrt the handling of unimplemented functionality and that in the next set of commits and subsequent new PR to account for unimplemented functionality more gracefully there.  With that said, I've removed the if placeholder logic for projection and replaced it with a single TODO comment.
https://github.com/cloudant/sync-android/commit/cf527fd540554a34338857f179c44fa358444ad3#diff-5f5fe154c6c97af9fc67a7b9f8ae9c87L143